### PR TITLE
RATIS-1641. Cleanup the Counter example and revise the "Getting Started" doc.

### PR DESCRIPTION
**Pull request by szetszwo**
See https://issues.apache.org/jira/browse/RATIS-1641

### CODE REVIEW

**Comment by szetszwo** [comment]
@drriguz , has you got a chance to check this?

**Comment by drriguz** [comment]
@szetszwo So sorry for not replying sooner! Too many big news those days...

The example is now very clear, and the document explains clearly about the key points when implement the statemachine :p

What I was thinking is very close to this, except that I was thinking extract some fileds/methods into a base class, eg. create a `SimpleStateMachine` class and owns a `SimpleStateMachineStorage`, but I did not find a good way to implement it.

Thanks for the excellent example! BTW, could you also invite me to the slack channel?

**Comment by drriguz** [comment]
BTW, I think we could also simplify the [ArithmeticStateMachine.java](https://github.com/apache/ratis/blob/master/ratis-examples/src/main/java/org/apache/ratis/examples/arithmetic/ArithmeticStateMachine.java) to use `synchronized` instead of `ReentrantReadWriteLock`, comparing with the benefit of performance, the code is less readable then the counter example.

**Comment by szetszwo** [comment]
> BTW, I think we could also simplify the [ArithmeticStateMachine.java](https://github.com/apache/ratis/blob/master/ratis-examples/src/main/java/org/apache/ratis/examples/arithmetic/ArithmeticStateMachine.java) to use synchronized instead of ReentrantReadWriteLock, comparing with the benefit of performance, the code is less readable then the counter example.

ArithmeticStateMachine indeed is a more complicated state machine than the CounterStateMachine.

Anyway, let's focus on the Counter example here.

> ... could you also invite me([dr@riguz.com](mailto:dr@riguz.com)) to the slack channel?

Just have invited you.  Please check.

File: ratis-examples/src/main/java/org/apache/ratis/examples/counter/server/CounterServer.java [hunk 1]
```diff
@@ -78,24 +78,41 @@ public void close() throws IOException {
...
   }
 
-  public static void main(String[] args) throws IOException {
-    if (args.length < 1) {
-      System.err.println("Usage: java -cp *.jar org.apache.ratis.examples.counter.server.CounterServer {serverIndex}");
-      System.err.println("{serverIndex} could be 1, 2 or 3");
+  public static void main(String[] args) {
+    try {
+      //get peerIndex from the arguments
+      if (args.length == 1) {
```
Comment (codings-dan) [inline ratis-examples/src/main/java/org/apache/ratis/examples/counter/server/CounterServer.java:84]: This judgment condition seems to be wrong，args.length != 1

**Review by codings-dan — COMMENTED** [review]
@szetszwo Thanks for working on this. This change is really good, it helps beginners to understand ratis faster, I left some comments for reference

File: ratis-docs/src/site/markdown/start/index.md [hunk 2]
```diff
@@ -43,13 +71,14 @@ it's available in maven central:
...
 
-Also, one of the following transports need to be added:
+Then, add one of the following transports:
 
-* grpc
-* netty
-* hadoop
+* ratis-grpc
+* ratis-netty
+* ratis-hadoop
```
[the remark this answers is not in the document]
Comment (szetszwo) [reply ratis-docs/src/site/markdown/start/index.md:78]: We still support ratis-hadoop although we did not update code recently.  The ratis-hadoop code is located in separated repo but the maven ratis-hadoop dependency still works.

File: ratis-docs/src/site/markdown/start/index.md [hunk 3]
```diff
@@ -61,294 +90,436 @@ For example, let's use grpc transport:
...
+   * @return the index of the snapshot
+   */
+  @Override
+  public long takeSnapshot() {
+    //get the current state
+    final CounterState state = getState();
+    final long index = state.getApplied().getIndex();
+
+    //create a file with a proper name to store the snapshot
+    final File snapshotFile = storage.getSnapshotFile(state.getApplied().getTerm(), index);
```
[the remark this answers is not in the document]
Comment (szetszwo) [reply ratis-docs/src/site/markdown/start/index.md:284]:
```
  public File getSnapshotFile(long term, long endIndex) {
    return new File(smDir, getSnapshotFileName(term, endIndex));
  }
```

getSnapshotFile creates a java object but not a real file in the local file system.   `createSnapshotFile` usually means creating a real file, e.g. java.nio.file.Files.createFile(..).

File: ratis-examples/src/main/java/org/apache/ratis/examples/counter/server/CounterServer.java (hunk 1 above)
Comment (szetszwo) [reply ratis-examples/src/main/java/org/apache/ratis/examples/counter/server/CounterServer.java:84]: Good catch!  Will update it.

File: ratis-docs/src/site/markdown/start/index.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (codings-dan) [reply ratis-docs/src/site/markdown/start/index.md:284]: Thanks for letting me know, `getSnapshotFile` is proper

**Review by codings-dan — APPROVED** [review]
The change looks good, +1

**Pull request merged**